### ISSUE TITLE
Allow jinja2 code in webhook payload_url

### ISSUE THREAD

**Issue by rodvand** [netbox-community/netbox#8295]
### NetBox version

v3.1.5

### Feature type

New functionality

### Proposed functionality

Allow the use of jinja2 code in the webhook payload_url.

### Use case

I have a custom field where I store the ID of an object in my NMS. This allows me to utilise NetBox to control my monitoring (pause monitoring if the device is set to offline as an example). The current setup requires some middleware where I POST to a generic URL and this web server again make a call to the REST API of the monitoring system using the ID in the custom field from the payload data. If I could reference the obj.cf_nms_id directly in the payload_url it would allow me to skip this middleware.

The payload_url would have to be generated on each webhook event.

### Database changes

None

**Comment by rodvand** [comment]
I've had a quick look in the code and you could probably extend the Webhook model with a method 
```
def render_payload_url(self, context):
        """
        Render the payload_url
        """
        return render_jinja2(self.payload_url, context)
```

and then reference it in https://github.com/netbox-community/netbox/blob/f94c1e91ea20c805528125407193371476ccbcd8/netbox/extras/webhooks_worker.py#L31 to get the generated URL.

If the FR is ok and accepted I can provide a PR for consideration.

**Comment by jeremystretch** [comment]
> The current setup requires some middleware where I POST to a generic URL and this web server again make a call to the REST API of the monitoring system using the ID in the custom field from the payload data.

Just as illustration for the use case, I take it you want to do something like this?

```
POST https://nms.local/devices/{{ object.cf.nms_id }}/edit/
```

Makes sense to me. My only concern is the potential for conflict with existing URLs if we enable Jinja2 rendering (for example, if someone has a URL defined that is evaluated as invalid Jinja2 code), although I'd expect such instances to be very rare.

**Comment by rodvand** [comment]
> > The current setup requires some middleware where I POST to a generic URL and this web server again make a call to the REST API of the monitoring system using the ID in the custom field from the payload data.
> 
> 
> 
> Just as illustration for the use case, I take it you want to do something like this?
> 
> 
> 
> ```
> 
> POST https://nms.local/devices/{{ object.cf.nms_id }}/edit/
> 
> ```
Yes, that's pretty much exactly my use case

> 
> 
> Makes sense to me. My only concern is the potential for conflict with existing URLs if we enable Jinja2 rendering (for example, if someone has a URL defined that is evaluated as invalid Jinja2 code), although I'd expect such instances to be very rare.

I thought about the same thing, but couldn't really wrap my head around how a URL like that would look.

**Comment by jeremystretch** [comment]
> I thought about the same thing, but couldn't really wrap my head around how a URL like that would look.

I've learned never to underestimate the bizarre ways in which some enterprise application will abuse URL query parameters. :laughing: But maybe it would be sufficient to gate this change in a milestone release (e.g. v3.2 or v3.3).

**Comment by geor-g** [comment]
This could potentially be gated via a config option, to lower potential risks, and make this opt-int; although I believe the risk is low in any case.

**Comment by rodvand** [comment]
> > I thought about the same thing, but couldn't really wrap my head around how a URL like that would look.
> 
> I've learned never to underestimate the bizarre ways in which some enterprise application will abuse URL query parameters. 😆 But maybe it would be sufficient to gate this change in a milestone release (e.g. v3.2 or v3.3).

I agree it would make sense to add this potential breaking change to a milestone release. If it breaks an URL it should just be some small adjustments to make it work with jinja2.

**Comment by jeremystretch** [comment]
@rodvand assigning this to you since you volunteered earlier. Just be sure to work off of the `feature` branch. (Or just let me know if you'd prefer I take it.) Thanks!
